### PR TITLE
Add support for loading segmentation datasets in Pascal VOC format

### PR DESCRIPTION
**Pull request by kirilllzaitsev**
# Description

Add possibility to load PASCAL VOC segmentation masks in addition to object detection ([related issue](https://github.com/roboflow/supervision/issues/244)).
Changes were made to core.DetectionDataset.from_pascal_voc and format.pascal_voc.load_pascal_voc_annotations.

## Type of change

Please delete options that are not relevant.

-   [+] New feature (non-breaking change which adds functionality)

## How has this change been tested, please provide a testcase or example of how you tested the change?

Via new tests.

## Any specific deployment considerations

For example, documentation changes, usability, usage/costs, secrets, etc.

## Docs

-   [ ] Docs updated? What were the changes:

### CODE REVIEW

File: supervision/dataset/formats/pascal_voc.py [hunk 1]
```diff
@@ -120,6 +124,99 @@ def detections_to_pascal_voc(
...
+    classes = list(set(classes))
+
+    for annotation in annotations.values():
+        class_id = [classes.index(class_name) for class_name in annotation.class_id]
+        annotation.class_id = np.array(class_id)
+
+    return classes, images, annotations
+
+
+def polygon_to_mask(polygon: np.ndarray, resolution_wh: Tuple[int, int]) -> np.ndarray:
```
Comment (mayankagarwals) [inline supervision/dataset/formats/pascal_voc.py:213]: There is `polygon_to_mask` available in detection.utils

File: supervision/dataset/core.py [hunk 2]
```diff
@@ -197,7 +197,10 @@ def as_pascal_voc(
 
     @classmethod
     def from_pascal_voc(
-        cls, images_directory_path: str, annotations_directory_path: str
+        cls,
+        images_directory_path: str,
+        annotations_directory_path: str,
+        force_masks: bool = False,
```
Comment (SkalskiP) [inline supervision/dataset/core.py:203]: Let's add `force_masks` into the docstring below.

File: supervision/dataset/core.py [hunk 3]
```diff
@@ -222,7 +225,7 @@ def from_pascal_voc(
             >>> project = rf.workspace(WORKSPACE_ID).project(PROJECT_ID)
             >>> dataset = project.version(PROJECT_VERSION).download("voc")
 
-            >>> ds = sv.DetectionDataset.from_yolo(
+            >>> ds = sv.DetectionDataset.from_pascal_voc(
```
Comment (SkalskiP) [inline supervision/dataset/core.py:228]: Awesome, you noticed and fixed that. 🙏🏻

File: supervision/dataset/core.py [hunk 4]
```diff
@@ -197,7 +197,10 @@ def as_pascal_voc(
 
     @classmethod
     def from_pascal_voc(
-        cls, images_directory_path: str, annotations_directory_path: str
+        cls,
```
Comment (SkalskiP) [inline supervision/dataset/core.py:200]: There is an unused import at the top of this file.

`from supervision.utils.file import list_files_with_extensions`

Please remove it if you can.

**Comment by SkalskiP** [comment]
> new implementation is not compatible with v1 of load_pascal_voc_annotations, but matches the loading standard of YOLO and COCO. Shall we temporarily leave the v1 there and issue a warning, or change the implementation to be backward compatible? Is there a general rule that we follow for such cases?

@kirilllzaitsev don't worry about breaking old `load_pascal_voc_annotations`. It is a function that is not publicly exposed.

File: supervision/dataset/formats/pascal_voc.py [hunk 5]
```diff
@@ -120,6 +124,93 @@ def detections_to_pascal_voc(
...
+
+                mask_from_polygon = polygon_to_mask(
+                    polygon=np.array(polygon_points),
+                    resolution_wh=(image.shape[0], image.shape[1]),
+                )
+                masks.append(mask_from_polygon)
+
+        xyxy = np.array(xyxy)
+        masks = np.array(masks)
+        annotation = Detections(xyxy=xyxy, mask=masks, class_id=np.array(class_names))
```
[the remark this answers is not in the document]
Comment (kirilllzaitsev) [reply supervision/dataset/formats/pascal_voc.py]: The initial motivation was to temporarily introduce this inconsistency until we can identify class_id by indexing all of the classes in annotations. But now it is implemented cleaner.

**Comment by kirilllzaitsev** [comment]
@SkalskiP , hi, ready for review.

**Comment by SkalskiP** [comment]
Hi, @kirilllzaitsev 👋🏻! Taking a look now.

**Comment by SkalskiP** [comment]
Hi, @kirilllzaitsev 👋🏻! I've done some tests, and I see some problems:

- `from_pascal_voc` loads `mask` data even if `force_masks=False.
- `mask.shape` have incorrect order of channels. It is `(N, W, H)`. It should be `(N, H, W)`.
- `mask.dtype` is incorrect. It is `uint8`. It should be `bool`.

Here is a Google Colab you can use to verify my findings: https://colab.research.google.com/drive/18rDUnAwPxhMt9YVbnFof5MS6__59NL11?usp=sharing

File: supervision/dataset/formats/pascal_voc.py [hunk 6]
```diff
@@ -120,6 +124,142 @@ def detections_to_pascal_voc(
...
+
+    if with_masks:
+        annotation = Detections(xyxy=xyxy, mask=np.array(masks), class_id=class_id)
+    else:
+        annotation = Detections(xyxy=xyxy, class_id=class_id)
+    return annotation, extended_classes
+
+
+def parse_polygon_points(polygon: Element) -> List[List[int]]:
+    # Parses polygon points in format: <polygon><x1>...</x1><y1>...</y1><x2>...</x2><y2>...</y2>...</polygon>
```
Comment (SkalskiP) [inline supervision/dataset/formats/pascal_voc.py:252]: I think we don't need that comment, as we have unit tests.

**Review by SkalskiP — CHANGES_REQUESTED** [review]

**Comment by SkalskiP** [comment]
@kirilllzaitsev let me know when you'll be ready ;)

**Comment by kirilllzaitsev** [comment]
> Hi, @kirilllzaitsev 👋🏻! I've done some tests, and I see some problems:
> 
> * `from_pascal_voc` loads `mask` data even if `force_masks=False`.
> * `mask.shape` have incorrect order of channels. It is `(N, W, H)`. It should be `(N, H, W)`.
> * `mask.dtype` is incorrect. It is `uint8`. It should be `bool`.
> 
> Here is a Google Colab you can use to verify my findings: https://colab.research.google.com/drive/18rDUnAwPxhMt9YVbnFof5MS6__59NL11?usp=sharing
> 
> I'm not sure yet what to do with `force_masks`. I see there is inconsistency across different format loaders.

To me, setting the force_masks flag translates into 'masks are required, and if there are no masks - raise an error'. This is what the YOLO loader does, contrary to the COCO that uses force_masks as an indicator of whether to use masks or not. The letter seems confusing to me.

**Comment by kirilllzaitsev** [comment]
@SkalskiP , ready

**Comment by SkalskiP** [comment]
> To me, setting the force_masks flag translates into 'masks are required, and if there are no masks - raise an error'. This is what the YOLO loader does, contrary to the COCO that uses force_masks as an indicator of whether to use masks or not. The latter seems confusing to me.

I think that our current API sucks. `force_masks` is completely non-intuitive. And of course, this is my fault. :) I just think about how to fix it.

Things we can do:
- Keep `force_masks` but make this behavior consistent for all formats.
- Have separate methods for detection and segmentation. For example `from_coco` and `from_coco_cegemtnation`.

@hardikdava what do you think?

**Comment by hardikdava** [comment]
@SkalskiP , I completely agree with you. We can introduce `sv.VisionTask` as argument to `sv.DetectionDataset` for scalable and specifying types which can be standardize.

```python
class VisionTask(Enum):
	CLASSIFICATION = 0
	OBJECT_DETECTION = 1
	ORIENTED_BOUNDING_BOX = 2
	INSTANCE_SEGMENTATION = 3
	KEYPOINTS_DETECTION = 4
	POSE_ESITMATION = 5
```
what do you think?

File: test/dataset/formats/test_pascal_voc.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+def test_object_to_pascal_voc(
+    xyxy: np.ndarray,
+    name: str,
+    polygon: Optional[np.ndarray],
+    expected_result,
+    exception: Exception,
+):
+    with exception:
+        result = object_to_pascal_voc(xyxy=xyxy, name=name, polygon=polygon)
+        with open("/tmp/test.xml", "w") as f:
```
Comment (SkalskiP) [inline test/dataset/formats/test_pascal_voc.py:64]: @kirilllzaitsev, what are we testing here? 👇🏻 

```python
with open("/tmp/test.xml", "w") as f:
    f.write(ET.tostring(result).decode())
with open("/tmp/exptest.xml", "w") as f:
    f.write(ET.tostring(expected_result).decode())
```

Comment (SkalskiP) [reply test/dataset/formats/test_pascal_voc.py:64]: To be honest, I'd drop those four lines.

File: test/dataset/formats/test_pascal_voc.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+            """<annotation><object><name>test</name><bndbox><xmin>0</xmin><ymin>0</ymin><xmax>10</xmax><ymax>10</ymax></bndbox></object></annotation>""",
+            ["test"],
+            (100, 100),
+            False,
+            mock_detections(np.array([[0, 0, 10, 10]]), None, [0]),
+            DoesNotRaise(),
+        )
+    ],
+)
+def test_detections_from_xml_obj(
```
Comment (SkalskiP) [inline test/dataset/formats/test_pascal_voc.py:106]: I'd love to see more tests:
- Empty annotation. 
- Annotation with multiple boxes of the same class.
- Annotation with multiple boxes of different classes.

File: supervision/dataset/formats/pascal_voc.py [hunk 9]
```diff
@@ -120,6 +124,141 @@ def detections_to_pascal_voc(
...
+    polygon_points = []
+    coords = polygon.findall(".//*")
+    for i in range(0, len(coords), 2):
+        x = int(coords[i].text)
+        y = int(coords[i + 1].text)
+        polygon_points.append([x, y])
+    return polygon_points
+
+
+def load_pascal_voc_annotations_v1(
```
Comment (SkalskiP) [inline supervision/dataset/formats/pascal_voc.py:261]: Let's drop that logic altogether.

**Review by SkalskiP — CHANGES_REQUESTED** [review]

**Comment by SkalskiP** [comment]
@kirilllzaitsev I left a few more final comments. But we are definitely on the right path.

**Review by SkalskiP — APPROVED** [review]

**Comment by SkalskiP** [comment]
Looks good to me. Merging! @kirilllzaitsev you plan to work on anything more in this release?

**Comment by kirilllzaitsev** [comment]
@SkalskiP sure, what else is on the plate?

**Comment by SkalskiP** [comment]
@kirilllzaitsev, how about splitting `DetectionDataset` into `DetectionDataset` and `SegmentationDataset` POC? https://github.com/roboflow/supervision/issues/244

**Pull request merged**